lib/rdf-query/src/matcher.rs

```diff
@@ -0,0 +1,56 @@
+extern crate alloc;
+
+use alloc::boxed::Box;
+use rdf_model::Term;
+
+use crate::variable::Variable;
+
+pub enum Matcher {
+    Variable(Variable),
+    Term(Box<dyn Term>),
+}
+
+impl Matcher {
+    pub fn as_variable(&self) -> Option<&Variable> {
+        match self {
+            Self::Variable(var) => Some(var),
+            _ => None,
+        }
+    }
+}
+
+impl PartialEq<&dyn Term> for Matcher {
+    fn eq(&self, term: &&dyn Term) -> bool {
+        match self {
+            Self::Variable(_) => true,
+            Self::Term(t) => t.as_str() == term.as_str(),
+        }
+    }
+}
+
+impl From<Variable> for Matcher {
+    fn from(var: Variable) -> Self {
+        Self::Variable(var)
+    }
+}
+
+impl From<&str> for Matcher {
+    fn from(name: &str) -> Self {
+        name.into()
+    }
+}
```

Comment:
This implementation is self-recursing.
```
warning: function cannot return without recursing
  --> lib/rdf-query/src/matcher.rs:38:5
   |
38 | /     fn from(name: &str) -> Self {
39 | |         name.into()
40 | |     }
   | |_____^
   |
note: recursive call site
  --> lib/rdf-query/src/matcher.rs:39:9
   |
39 |         name.into()
   |         ^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unconditional_recursion
   = note: `#[warn(clippy::unconditional_recursion)]` on by default
```

```suggestion
impl From<&str> for Matcher {
    fn from(name: &str) -> Self {
        Variable::from(name).into()
    }
}
```